File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = prev = ListNode()
        curr1 = list1
        curr2 = list2

        while curr1 and curr2:
            if curr1.val < curr2.val:
                # temp = curr1.next
                # curr1.next = None
                # prev.next = curr1
                # curr1 = temp
                prev.next = curr1
                curr1 = curr1.next
            else:
                prev.next = curr2
                curr2 = curr2.next
                # temp = curr2.next
                # curr2.next = None
                # prev.next = curr2
                # curr2 = temp
            prev = prev.next
    
        if curr1:
            prev.next = curr1
        else:
            prev.next = curr2    


        return dummy.next
        

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        //(0) - [1,2,3,4],
        ListNode fast = head;
        ListNode dummy = new ListNode(0, head);
        ListNode slow = dummy;
        while(n > 0) {
            fast = fast.next;
            n--;
        }

        while(fast != null) {
            slow = slow.next;
            fast = fast.next;
        }

        slow.next = slow.next.next;

        return dummy.next;
    }
}
